--- trunk~/src/luaconf.h	2025-10-07 22:00:10.960008872 -0400
+++ trunk/src/luaconf.h	2025-10-07 22:00:31.934916268 -0400
@@ -35,8 +35,8 @@
 
 #if defined(LUA_USE_LINUX)
 #define LUA_USE_POSIX
-#define LUA_USE_DLOPEN		/* needs an extra library: -ldl */
-#define LUA_USE_READLINE	/* needs some extra libraries */
+// #define LUA_USE_DLOPEN		/* needs an extra library: -ldl */
+// #define LUA_USE_READLINE	/* needs some extra libraries */
 #endif
 
 #if defined(LUA_USE_MACOSX)